gk/main.c

```diff
@@ -686,6 +797,321 @@ gk_conf_put(struct gk_config *gk_conf)
+{
+	int i;
+	int ret = 0;
+	int num_ipv4_routes = 0;
+	int num_ipv6_routes = 0;
+	struct ipv4_lpm_route *ipv4_routes;
+	struct ipv6_lpm_route *ipv6_routes;
+
+	gk_conf->net = net_conf;
+
+	if (gk_conf->num_lcores <= 0)
+		goto out;
+
+	/*
+	 * Set up the GK LPM table. We assume that
+	 * all the GK instances are running on the same socket.
+	 */
+	ret = setup_gk_lpm(gk_conf,
+		rte_lcore_to_socket_id(gk_conf->lcores[0]));
+	if (ret < 0)
+		goto out;
+
+	for (i = 0; i < GK_MAX_NUM_POLICIES; i++)
+		rte_atomic16_init(&gk_conf->lpm_tbl.policy_tbl[i].ref_cnt);
+
+	ipv4_routes = rte_calloc(NULL,
+		gk_conf->max_num_ipv4_rules,
+		sizeof(*ipv4_routes), 0);
+	if (ipv4_routes == NULL) {
+		ret = -1;
+		goto free_lpm;
+	}
+
+	ipv6_routes = rte_calloc(NULL,
+		gk_conf->max_num_ipv6_rules,
+		sizeof(*ipv6_routes), 0);
+	if (ipv6_routes == NULL) {
+		ret = -1;
+		goto free_ipv4_routes;
+	}
+
+	/* Parse all the routes. */
+	for (i = 0; i < num_routes; i++) {
+		struct in_addr ipv4_addr;
+		struct in6_addr ipv6_addr;
+		int ip_type = get_ip_type(routes[i].ip_addr);
+		if (ip_type == AF_INET) {
+			if (inet_pton(AF_INET,
+					routes[i].ip_addr, &ipv4_addr) != 1) {
+				ret = -1;
+				goto free_ipv6_routes;
+			}
+
+			ipv4_routes[num_ipv4_routes].ip =
+				ipv4_addr.s_addr;
+			ipv4_routes[num_ipv4_routes].depth =
+				routes[i].prefix_len;
+			ipv4_routes[num_ipv4_routes].policy_id =
+				routes[i].policy_id;
+			num_ipv4_routes++;
+		} else if (ip_type == AF_INET6) {
+			if (inet_pton(AF_INET6,
+					routes[i].ip_addr, &ipv6_addr) != 1) {
+				ret = -1;
+				goto free_ipv6_routes;
+			}
+
+			rte_memcpy(ipv6_routes[num_ipv6_routes].ip,
+				ipv6_addr.s6_addr,
+				sizeof(ipv6_routes[num_ipv6_routes].ip));
+			ipv6_routes[num_ipv6_routes].depth =
+				routes[i].prefix_len;
+			ipv6_routes[num_ipv6_routes].policy_id =
+				routes[i].policy_id;
+			num_ipv6_routes++;
+		} else {
+			ret = -1;
+			goto free_ipv6_routes;
+		}
+	}
+
+	/* Add IP routes to the LPM table. */
+
+	/* Set up the IPv4 routes. */
+	ret = gk_lpm_add_ipv4_routes(
+		gk_conf->lpm_tbl.lpm, ipv4_routes,
+		num_ipv4_routes, &gk_conf->lpm_tbl);
+	if (ret < 0)
+		goto free_ipv6_routes;
+
+	/* Set up the IPv6 routes. */
+	ret = gk_lpm_add_ipv6_routes(
+		gk_conf->lpm_tbl.lpm6, ipv6_routes,
+		num_ipv6_routes, &gk_conf->lpm_tbl);
+	if (ret < 0)
+		goto free_ipv6_routes;
+
+	/* Add policies to the routing table. */
+	ret = add_lpm_policy(&gk_conf->lpm_tbl, policies, num_policies);
+	if (ret < 0)
+		goto free_ipv6_routes;
+
+	/* Initialize the tunnels' information. */
+	ret = lua_init_tunnels(gk_conf, grantor_addrs, num_grantors);
+	if (ret < 0)
+		goto free_ipv6_routes;
+
+	/* Before exiting this function, needs to free the routes. */
+	rte_free(ipv4_routes);
+	rte_free(ipv6_routes);
+
+	ret = 0;
+	goto out;
+
+free_ipv6_routes:
+	rte_free(ipv6_routes);
+
+free_ipv4_routes:
+	rte_free(ipv4_routes);
+
+free_lpm:
+	destroy_gk_lpm(&gk_conf->lpm_tbl);
+
+	rte_free(gk_conf);
```

Comment:
We don't free `gk_conf` in `run_gk()` when initialization fails -- is there a good reason to do it here?